pecan/lang/pecan_ast/arith.py

```diff
@@ -140,39 +103,48 @@ def evaluate(self, prog):
     def __repr__(self):
         return '({} * {})'.format(self.a, self.b)
 
+    def evaluate_int(self, prog):
+        assert self.is_int
+        return self.a.evaluate_int(prog) * self.b.evaluate_int(prog)
 #TODO:
 class Div(BinaryExpression):
     def __init__(self, a, b):
         super().__init__(a, b)
         self.a = a
         self.b = b
-        raise NotImplementedError("Division hasn't been implemented, sorry. {}".format(self))
+        if not self.is_int:
+            raise NotImplementedError("Division with automaton hasn't been implemented, sorry. {}".format(self))
+        if not self.b.is_int:
+            raise AutomatonArithmeticError("Second argument of division must be an integer in {}".format(self))
 
     def __repr__(self):
         return '({} / {})'.format(self.a, self.b)
 
     def evaluate(self, prog):
-        if type(self.b) is not IntConst:
-            raise AutomatonArithmeticError("Second argument of division must be an integer in {}".format(self))
-        if type(self.a) is IntConst:
-            if self.a % self.b != 0:
-                raise AutomatonArithmeticError("Division among integers must output an integer in {}".format(self))
-            return self.a // self.b
-        if self.b == 1:
+        if self.is_int:
+            return IntConst(self.evaluate_int(prog))
+        assert False 
+        b = self.b.evaluate_int(prog)
+        if b == 1:
             return self.a.evaluate(prog)
 
         (aut_a, val_a) = self.a.evaluate(prog)
-        (aut_b, val_b) = self.b.evaluate(prog)
+        # (aut_b, val_b) = IntConst(b).evaluate(prog)
         #TODO: change label, not finished
         (aut_div,val_div) = Mul(self.b,spot.formula('1').translate()).evaluate(prog)
         def build_div_formula(formula):
-            return Substitution({val_div: spot.formula(val_a), 'a': spot.formula('{}_div_{}'.format(val_a,val_b))}).substitute(formula)
+            return Substitution({val_div: spot.formula(val_a), 'a': spot.formula('{}_div_{}'.format(val_a,val_a))}).substitute(formula)
         #TODO: drop val_a, val_b in return
-        return (spot.product(aut_a, spot.product(aut_b, aut_div)), '{}_add_{}'.format(self.a,self.b))
+        return (spot.product(aut_a, spot.product(aut_a, aut_div)), '{}_add_{}'.format(self.a,self.b))
```

Comment:
Why are there two `aut_a`'s here?

Reply:
Because I thought I would get a compiler error if I use the undeclared aut_b. But I forgot I'm in Python.